File: SOLID/code-examples/chapter-04.cs
// FILE: chapter-04-basic.cs
// Chapter 4 — Inheritance — Basic

using System;

// Base class (parent) - simple animal with name
public class Animal
{
    public string Name;
    
    public Animal(string name)
    {
        Name = name;
    }
    
    public void Eat()
    {
        Console.WriteLine($"{Name} is eating");
    }
    
    public virtual void MakeSound()
    {
        Console.WriteLine($"{Name} makes a sound");
    }
}

// Derived class (child) - inherits from Animal
public class Dog : Animal
{
    public Dog(string name) : base(name)
    {
    }
    
    // Override parent method
    public override void MakeSound()
    {
        Console.WriteLine($"{Name} says: Woof!");
    }
    
    // Dog-specific behavior
    public void WagTail()
    {
        Console.WriteLine($"{Name} wags tail");
    }
}

class Program
{
    static void Main()
    {
        Console.WriteLine("🐕 Basic Inheritance Example");
        Console.WriteLine("===========================\n");
        
        // Create a dog
        Dog myDog = new Dog("Buddy");
        
        // Use inherited method
        myDog.Eat();        // From Animal class
        
        // Use overridden method
        myDog.MakeSound();  // Dog's version
        
        // Use dog-specific method
        myDog.WagTail();    // Only dogs can do this
        
        Console.WriteLine("\n💡 Inheritance Benefits:");
        Console.WriteLine("   ✓ Code reuse: Dog inherits Eat() from Animal");
        Console.WriteLine("   ✓ Customization: Dog overrides MakeSound()");
        Console.WriteLine("   ✓ Extension: Dog adds WagTail() behavior");
    }
}

File: SOLID/code-examples/chapter-04.cpp
// FILE: chapter-04.cpp
// Chapter 4 — Inheritance — C++

#include <iostream>
#include <string>
using namespace std;

// Base class (parent)
class Animal {
protected:
    string name;
    
public:
    Animal(const string& name) : name(name) {}
    
    void eat() {
        cout << name << " is eating" << endl;
    }
    
    virtual void makeSound() {
        cout << name << " makes a sound" << endl;
    }
    
    virtual ~Animal() = default; // Virtual destructor
};

// Derived class (child)
class Dog : public Animal {
public:
    Dog(const string& name) : Animal(name) {}
    
    // Override parent method
    void makeSound() override {
        cout << name << " says: Woof!" << endl;
    }
    
    // Dog-specific behavior
    void wagTail() {
        cout << name << " wags tail" << endl;
    }
};

int main() {
    cout << "🐕 Basic Inheritance Example (C++)" << endl;
    cout << "===================================" << endl << endl;
    
    // Create a dog
    Dog myDog("Buddy");
    
    // Use inherited method
    myDog.eat();        // From Animal class
    
    // Use overridden method
    myDog.makeSound();  // Dog's version
    
    // Use dog-specific method
    myDog.wagTail();    // Only dogs can do this
    
    cout << endl << "💡 Inheritance Benefits:" << endl;
    cout << "   ✓ Code reuse: Dog inherits eat() from Animal" << endl;
    cout << "   ✓ Customization: Dog overrides makeSound()" << endl;
    cout << "   ✓ Extension: Dog adds wagTail() behavior" << endl;
    
    return 0;
}